platform/core/src/services/DisplaySetService/DisplaySetService.ts

```diff
@@ -236,34 +295,81 @@ export default class DisplaySetService extends PubSubService {
+              this.activeDisplaySetsChanged = true;
+              instances = filterInstances(instances, [addedDs]);
+              this._addActiveDisplaySets([addedDs]);
+              this.setDisplaySetMetadataInvalidated(
+                addedDs.displaySetInstanceUID
+              );
+            }
+            // This means that all instances already existed or got added to
+            // existing display sets, and had an invalidated event fired
+            if (!instances.length) return allDisplaySets;
+          }
+
+          if (!instances.length) {
+            // Everything is already added - this is just an update caused
+            // by something else
+            this._addActiveDisplaySets(displaySets);
+            return allDisplaySets;
+          }
+        }
 
-          if (!displaySets || !displaySets.length) continue;
+        // The instances array still contains some instances, so try
+        // creating additional display sets using the sop class handler
+        displaySets = handler.getDisplaySetsFromSeries(instances);
 
-          // applying hp-defined viewport settings to the displaysets
-          displaySets.forEach(ds => {
-            Object.keys(settings).forEach(key => {
-              ds[key] = settings[key];
-            });
-          });
+        if (!displaySets || !displaySets.length) continue;
 
-          this._addDisplaySetsToCache(displaySets);
-          this._addActiveDisplaySets(displaySets);
+        // applying hp-defined viewport settings to the displaysets
+        displaySets.forEach(ds => {
+          Object.keys(settings).forEach(key => {
+            ds[key] = settings[key];
+          });
+        });
 
-          instances = instances.filter(
-            instance => !findInstance(instance, displaySets)
-          );
-        }
+        this._addDisplaySetsToCache(displaySets);
+        this._addActiveDisplaySets(displaySets);
 
-        allDisplaySets = allDisplaySets
-          ? [...allDisplaySets, ...displaySets]
-          : displaySets;
+        // It is possible that this SOP class handler handled some instances
+        // but there may need to be other instances handled by other handlers,
+        // so remove the handled instances
+        instances = filterInstances(instances, displaySets);
 
-        if (!instances.length) return allDisplaySets;
+        allDisplaySets.push(...displaySets);
       }
     }
     return allDisplaySets;
   }
+
+  /**
+   * Iterates over displaysets and invokes comparator for each element.
+   * It returns a list of items that has being succeed by comparator method.
+   *
+   * @param {function} comparator method to be used on the validation
+   * @returns list of displaysets
+   */
+  getDisplaySetsBy(comparator) {
+    const result = [];
+
+    if (typeof comparator !== 'function') {
+      throw new Error(`The comparator ${comparator} was not a function`);
+    }
+
+    this.getActiveDisplaySets().forEach(displaySet => {
```

Comment:
Why not use `.filter` here?

Reply:
In their infinite wisdom, the JS gods decided that iterators don't have the filter method :-(